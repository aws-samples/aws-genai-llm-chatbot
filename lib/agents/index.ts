import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as iam from "aws-cdk-lib/aws-iam";
import * as path from "path";
import * as geo from "aws-cdk-lib/aws-location";
import { bedrock } from "@cdklabs/generative-ai-cdk-constructs";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";
import { VectorIndex } from "@cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/opensearch-vectorindex";
import { VectorCollection } from "@cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/opensearchserverless";
import { NagSuppressions } from "cdk-nag";

export class BedrockWeatherAgent extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const powertools = lambda.LayerVersion.fromLayerVersionArn(
      this,
      "powertools",
      `arn:aws:lambda:${
        cdk.Stack.of(this).region
      }:017000801446:layer:AWSLambdaPowertoolsPythonV2:58`
    );

    const placeIndex = new geo.CfnPlaceIndex(this, "place-index", {
      dataSource: "Esri",
      indexName: "PlaceIndex",
      pricingPlan: "RequestBasedUsage",
    });

    const vectorStore = new VectorCollection(this, "vector-collection");
    const vectorIndex = new VectorIndex(this, "vector-index", {
      collection: vectorStore,
      indexName: "weather",
      vectorField: "vector",
      vectorDimensions: 1536,
      mappings: [
        {
          mappingField: "AMAZON_BEDROCK_TEXT_CHUNK",
          dataType: "text",
          filterable: true,
        },
        {
          mappingField: "AMAZON_BEDROCK_METADATA",
          dataType: "text",
          filterable: false,
        },
      ],
    });

    const modulesLayer = new lambda.LayerVersion(this, "modules-layer", {
      code: lambda.Code.fromDockerBuild(
        path.join(__dirname, "weather/modules")
      ),
      compatibleRuntimes: [lambda.Runtime.PYTHON_3_12],
      description: "Layer with required modules for the agent",
    });

    const weather = new lambda.Function(this, "weather", {
      runtime: lambda.Runtime.PYTHON_3_12,
      description:
        "Lambda function that implements APIs to retrieve weather data",
      code: lambda.Code.fromAsset(path.join(__dirname, "weather")),
      handler: "lambda.handler",
      environment: {
        LOCATION_INDEX: placeIndex.indexName,
      },
      memorySize: 512,
      timeout: cdk.Duration.seconds(10),
      layers: [powertools, modulesLayer],
    });

    const policy = new iam.Policy(this, "weather-policy", {
      statements: [
        new iam.PolicyStatement({
          actions: ["geo:SearchPlaceIndexForText"],
          resources: ["*"],
        }),
      ],
    });
    weather.role?.attachInlinePolicy(policy);

    const kb = new bedrock.KnowledgeBase(this, "weather-kb", {
      embeddingsModel: bedrock.BedrockFoundationModel.TITAN_EMBED_TEXT_V1,
      vectorField: "vector",
      vectorIndex: vectorIndex,
      vectorStore: vectorStore,
      indexName: "weather",
      instruction: "answers questions about WMO and metereology",
    });

    const bucket = new Bucket(this, "bedrock-kb-datasource-bucket", {
      enforceSSL: true,
    });

    const keyPrefix = "my-docs";
    new bedrock.S3DataSource(this, "my-docs-datasource", {
      bucket: bucket,
      dataSourceName: "my-docs",
      knowledgeBase: kb,
      inclusionPrefixes: [keyPrefix],
    });

    const agent = new bedrock.Agent(this, "weather-agent", {
      foundationModel:
        bedrock.BedrockFoundationModel.ANTHROPIC_CLAUDE_INSTANT_V1_2,
      instruction: `You are a weather expert and answer user question about weather in different places. 
If you are asked to provide historical date, you need to answer with a summary of the weather over the period. 
You answer the questions in the same language they have been asked.`,
      description: "an agent to interact with a weather api",
      knowledgeBases: [kb],
      name: "WeatherTeller",
    });

    agent.addActionGroup({
      actionGroupExecutor: weather,
      apiSchema: bedrock.ApiSchema.fromAsset(
        path.join(__dirname, "weather", "schema.json")
      ),
      actionGroupState: "ENABLED",
    });

    agent.addActionGroup({
      parentActionGroupSignature: "AMAZON.UserInput",
      actionGroupState: "ENABLED",
      actionGroupName: "UserInputAction",
    });

    new BucketDeployment(this, "my-docs-files", {
      destinationBucket: bucket,
      destinationKeyPrefix: keyPrefix,
      sources: [Source.asset(path.join(__dirname, "my-documents"))],
    });

    new cdk.CfnOutput(this, "weather-function-name", {
      value: weather.functionName,
    });

    NagSuppressions.addResourceSuppressions(weather.role!, [
      {
        id: "AwsSolutions-IAM4",
        reason: "IAM role implicitly created by CDK.",
      },
    ]);

    NagSuppressions.addResourceSuppressions(policy, [
      {
        id: "AwsSolutions-IAM5",
        reason: "IAM role implicitly created by CDK.",
      },
    ]);

    NagSuppressions.addResourceSuppressions(bucket, [
      {
        id: "AwsSolutions-S1",
        reason: "Access logs not required",
      },
    ]);
  }
}
