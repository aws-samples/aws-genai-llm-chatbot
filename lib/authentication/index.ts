import * as cognitoIdentityPool from "@aws-cdk/aws-cognito-identitypool-alpha";
import * as cdk from "aws-cdk-lib";
import * as cognito from "aws-cdk-lib/aws-cognito";
import { Construct } from "constructs";

export class Authentication extends Construct {
  public readonly userPool: cognito.UserPool;
  public readonly userPoolClient: cognito.UserPoolClient;
  public readonly identityPool: cognitoIdentityPool.IdentityPool;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const userPool = new cognito.UserPool(this, "UserPool", {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      selfSignUpEnabled: false,
      autoVerify: { email: true, phone: true },
      signInAliases: {
        email: true,
      },
    });

    const userPoolClient = userPool.addClient("UserPoolClient", {
      generateSecret: false,
      authFlows: {
        adminUserPassword: true,
        userPassword: true,
        userSrp: true,
      },
    });

    const identityPool = new cognitoIdentityPool.IdentityPool(
      this,
      "IdentityPool",
      {
        authenticationProviders: {
          userPools: [
            new cognitoIdentityPool.UserPoolAuthenticationProvider({
              userPool,
              userPoolClient,
            }),
          ],
        },
      }
    );

    this.userPool = userPool;
    this.userPoolClient = userPoolClient;
    this.identityPool = identityPool;

    new cdk.CfnOutput(this, "UserPoolId", {
      value: userPool.userPoolId,
    });

    new cdk.CfnOutput(this, "UserPoolWebClientId", {
      value: userPoolClient.userPoolClientId,
    });

    new cdk.CfnOutput(this, "UserPoolLink", {
      value: `https://${
        cdk.Stack.of(this).region
      }.console.aws.amazon.com/cognito/v2/idp/user-pools/${
        userPool.userPoolId
      }/users?region=${cdk.Stack.of(this).region}`,
    });
  }
}
