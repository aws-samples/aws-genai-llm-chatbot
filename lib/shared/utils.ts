import * as fs from "node:fs";
import * as path from "node:path";
import { SystemConfig } from "./types";

export abstract class Utils {
  static copyDirRecursive(sourceDir: string, targetDir: string): void {
    if (!fs.existsSync(targetDir)) {
      fs.mkdirSync(targetDir);
    }

    const files = fs.readdirSync(sourceDir);

    for (const file of files) {
      const sourceFilePath = path.join(sourceDir, file);
      const targetFilePath = path.join(targetDir, file);
      const stats = fs.statSync(sourceFilePath);

      if (stats.isDirectory()) {
        Utils.copyDirRecursive(sourceFilePath, targetFilePath);
      } else {
        fs.copyFileSync(sourceFilePath, targetFilePath);
      }
    }
  }

  static getDefaultEmbeddingsModel(config: SystemConfig): string {
    const defaultModel = config.rag.embeddingsModels.find(
      (model) => model.default === true
    );

    if (!defaultModel) {
      throw new Error("No default embeddings model found");
    }

    return `${defaultModel.provider}::${defaultModel.dimensions}::${defaultModel.name}`;
  }

  static getDefaultCrossEncoderModel(config: SystemConfig): string {
    const defaultModel = config.rag.crossEncoderModels.find(
      (model) => model.default === true
    );

    if (!defaultModel) {
      throw new Error("No default cross encoder model found");
    }

    return `${defaultModel.provider}::${defaultModel.name}`;
  }

  static getName(config: SystemConfig, value: string): string {
    const prefix = config.prefix;
    let name = prefix && prefix.length > 0 ? `${prefix}-${value}` : value;

    // Convert name to lowercase
    name = name.toLowerCase();

    // Remove any characters that are not a-z, 0-9, or hyphen
    name = name.replace(/[^a-z0-9-]/g, '');

    // Ensure name doesn't start or end with a hyphen
    if (name.startsWith('-')) {
        name = name.slice(1);
    }
    if (name.endsWith('-')) {
        name = name.slice(0, -1);
    }

    // Ensure name is between 3 and 28 characters
    if (name.length < 3) {
        throw new Error('Generated name is too short. It must be between 3 and 28 characters.');
    }
    name = name.slice(0, 28);

    // Ensure name starts with a lowercase letter
    if (!/^[a-z]/.test(name)) {
        throw new Error('Generated name must start with a lowercase letter.');
    }

    return name;
  }
}
