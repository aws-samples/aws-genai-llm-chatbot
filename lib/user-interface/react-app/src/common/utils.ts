export class Utils {
  static isDevelopment() {
    return import.meta.env.MODE === "development";
  }

  // eslint-disable-next-line @typescript-eslint/ban-types
  static isFunction(value: unknown): value is Function {
    return typeof value === "function";
  }

  static classNames(...classes: string[]) {
    return classes.filter(Boolean).join(" ");
  }

  static generateUUID() {
    if (crypto && crypto.randomUUID) {
      return crypto.randomUUID();
    }

    if (crypto && crypto.getRandomValues) {
      console.log(
        "crypto.randomUUID is not available using crypto.getRandomValues"
      );

      return ("" + [1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(
        /[018]/g,
        (ch) => {
          const c = Number(ch);
          return (
            c ^
            (crypto.getRandomValues(new Uint8Array(1))[0] & (15 >> (c / 4)))
          ).toString(16);
        }
      );
    }

    console.log("crypto is not available");
    let date1 = new Date().getTime();
    let date2 =
      (typeof performance !== "undefined" &&
        performance.now &&
        performance.now() * 1000) ||
      0;

    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(
      /[xy]/g,
      function (c) {
        let r = Math.random() * 16;
        if (date1 > 0) {
          r = (date1 + r) % 16 | 0;
          date1 = Math.floor(date1 / 16);
        } else {
          r = (date2 + r) % 16 | 0;
          date2 = Math.floor(date2 / 16);
        }

        return (c === "x" ? r : (r & 0x3) | 0x8).toString(16);
      }
    );
  }

  static delay(delay: number) {
    return new Promise((resolve) => {
      setTimeout(resolve, delay);
    });
  }

  static findElementInParents(element: HTMLElement | null, tagName: string) {
    let current: HTMLElement | null = element;
    while (current) {
      if (current.tagName === tagName) {
        return current;
      }

      current = current.parentElement;
    }

    return null;
  }

  static getErrorMessage(error: any) {
    if (error.errors) {
      return error.errors.map((e: any) => e.message).join(", ");
    }

    return "Unknown error";
  }

  static urlSearchParamsToRecord(
    params: URLSearchParams
  ): Record<string, string> {
    const record: Record<string, string> = {};

    for (const [key, value] of params.entries()) {
      record[key] = value;
    }

    return record;
  }

  static bytesToSize(bytes: number): string {
    const sizes: string[] = ["Bytes", "KB", "MB", "GB", "TB", "PB"];

    if (bytes === 0) return "0 MB";
    const i: number = parseInt(
      Math.floor(Math.log(bytes) / Math.log(1024)).toString()
    );

    const sizeStr = i >= sizes.length ? "" : sizes[i];
    return Math.round(bytes / Math.pow(1024, i)) + " " + sizeStr;
  }

  static textEllipsis(str: string, maxLength: number): string {
    if (str.length <= maxLength) return str;
    return str.substring(0, maxLength - 3) + "...";
  }

  static isValidURL(value: string) {
    if (value.length === 0 || value.indexOf(" ") !== -1) {
      return false;
    }

    const result = value.match(
      /(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]+\.[^\s]{2,}|www\.[a-zA-Z0-9]+\.[^\s]{2,})/gi
    );

    return result !== null;
  }
}
