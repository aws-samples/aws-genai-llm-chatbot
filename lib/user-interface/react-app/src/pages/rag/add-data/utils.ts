import { SelectOption, ContentTypesOptionsConfig } from "./types";

export function generateSelectedOptions(
  contentTypes: (string | undefined)[]
): SelectOption[] {
  return contentTypes.map(
    (ct): SelectOption => ({
      label: ct || "text/html",
      value: ct || "text/html",
      description:
        ContentTypesOptionsConfig[ct as keyof typeof ContentTypesOptionsConfig]
          ?.description || "Default Description",
    })
  );
}
