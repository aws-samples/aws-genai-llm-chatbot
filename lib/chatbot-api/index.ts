import * as apigwv2 from "@aws-cdk/aws-apigatewayv2-alpha";
import * as apigateway from "aws-cdk-lib/aws-apigateway";
import * as cognito from "aws-cdk-lib/aws-cognito";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as sns from "aws-cdk-lib/aws-sns";
import * as ssm from "aws-cdk-lib/aws-ssm";
import { Construct } from "constructs";
import { RagEngines } from "../rag-engines";
import { Shared } from "../shared";
import { SageMakerModelEndpoint, SystemConfig } from "../shared/types";
import { ChatBotDynamoDBTables } from "./chatbot-dynamodb-tables";
import { ChatBotS3Buckets } from "./chatbot-s3-buckets";
import { RestApi } from "./rest-api";
import { WebSocketApi } from "./websocket-api";

export interface ChatBotApiProps {
  readonly shared: Shared;
  readonly config: SystemConfig;
  readonly ragEngines?: RagEngines;
  readonly userPool: cognito.UserPool;
  readonly modelsParameter: ssm.StringParameter;
  readonly models: SageMakerModelEndpoint[];
}

export class ChatBotApi extends Construct {
  public readonly restApi: apigateway.RestApi;
  public readonly webSocketApi: apigwv2.WebSocketApi;
  public readonly messagesTopic: sns.Topic;
  public readonly sessionsTable: dynamodb.Table;
  public readonly byUserIdIndex: string;
  public readonly filesBucket: s3.Bucket;

  constructor(scope: Construct, id: string, props: ChatBotApiProps) {
    super(scope, id);

    const chatTables = new ChatBotDynamoDBTables(this, "ChatDynamoDBTables");
    const chatBuckets = new ChatBotS3Buckets(this, "ChatBuckets");

    const restApi = new RestApi(this, "RestApi", {
      ...props,
      sessionsTable: chatTables.sessionsTable,
      byUserIdIndex: chatTables.byUserIdIndex,
    });

    const webSocketApi = new WebSocketApi(this, "WebSocketApi", props);

    this.restApi = restApi.api;
    this.webSocketApi = webSocketApi.api;
    this.messagesTopic = webSocketApi.messagesTopic;
    this.sessionsTable = chatTables.sessionsTable;
    this.byUserIdIndex = chatTables.byUserIdIndex;
    this.filesBucket = chatBuckets.filesBucket;
  }
}
