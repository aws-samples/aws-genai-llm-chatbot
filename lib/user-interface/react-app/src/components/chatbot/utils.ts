import { Storage } from "aws-amplify";
import { Dispatch, SetStateAction } from "react";
import {
  ChatBotAction,
  ChatBotHistoryItem,
  ChatBotMessageResponse,
  ChatBotMessageType,
  ChatInputState,
} from "./types";
import { ChatSession } from "./multi-chat";
import { ModelItem } from "../../common/types";
import { SelectProps } from "@cloudscape-design/components";
import { OptionsHelper } from "../../common/helpers/options-helper";

export function updateMessageHistory(
  sessionId: string,
  messageHistory: ChatBotHistoryItem[],
  setMessageHistory: Dispatch<SetStateAction<ChatBotHistoryItem[]>>,
  response: ChatBotMessageResponse,
  setState: Dispatch<SetStateAction<ChatInputState>>
) {
  if (response.data?.sessionId !== sessionId) return;

  if (
    response.action === ChatBotAction.FinalResponse ||
    response.action === ChatBotAction.Error
  ) {
    setState((state) => ({ ...state, running: false }));
  }
  if (
    response.action === ChatBotAction.LLMNewToken ||
    response.action === ChatBotAction.FinalResponse ||
    response.action === ChatBotAction.Error
  ) {
    const content = response.data?.content;
    let metadata = response.data?.metadata;
    const token = response.data?.token;
    const hasContent = typeof content !== "undefined";
    const hasToken = typeof token !== "undefined";
    const hasMetadata = typeof metadata !== "undefined";

    if (
      messageHistory.length > 0 &&
      messageHistory[messageHistory.length - 1]["type"] !==
        ChatBotMessageType.Human
    ) {
      const lastMessage = messageHistory[messageHistory.length - 1];
      lastMessage.tokens = lastMessage.tokens || [];
      if (hasToken) {
        lastMessage.tokens.push(token);
      }

      lastMessage.tokens.sort((a, b) => a.sequenceNumber - b.sequenceNumber);
      if (lastMessage.tokens.length > 0) {
        const lastRunId =
          lastMessage.tokens[lastMessage.tokens.length - 1].runId;
        if (lastRunId) {
          lastMessage.tokens = lastMessage.tokens.filter(
            (c) => c.runId === lastRunId
          );
        }
      }

      if (!hasMetadata) {
        metadata = lastMessage.metadata;
      }

      if (hasContent) {
        setMessageHistory((history) => [
          ...history.slice(0, history.length - 1),
          {
            ...lastMessage,
            type: ChatBotMessageType.AI,
            content,
            metadata,
            tokens: lastMessage.tokens,
          },
        ]);
      } else {
        const contentFromTokens = lastMessage.tokens
          .map((c) => c.value)
          .join("");

        setMessageHistory((history) => [
          ...history.slice(0, history.length - 1),
          {
            ...lastMessage,
            type: ChatBotMessageType.AI,
            content: contentFromTokens,
            metadata,
            tokens: lastMessage.tokens,
          },
        ]);
      }
    } else {
      if (hasContent) {
        const tokens = hasToken ? [token] : [];
        setMessageHistory((history) => [
          ...history,
          {
            type: ChatBotMessageType.AI,
            content,
            metadata,
            tokens,
          },
        ]);
      } else if (typeof token !== "undefined") {
        setMessageHistory((history) => [
          ...history,
          {
            type: ChatBotMessageType.AI,
            content: token.value,
            metadata,
            tokens: [token],
          },
        ]);
      }
    }
  }
}

export function updateChatSessions(
  chatSession: ChatSession,
  response: ChatBotMessageResponse,
): void {
  if (response.data?.sessionId !== chatSession.id) return;

  const messageHistory = chatSession.messageHistory;
  console.log(messageHistory)
  if (
    response.action === ChatBotAction.FinalResponse ||
    response.action === ChatBotAction.Error
  ) {
    chatSession.running = false;
  }
  if (
    response.action === ChatBotAction.LLMNewToken ||
    response.action === ChatBotAction.FinalResponse ||
    response.action === ChatBotAction.Error
  ) {
    const content = response.data?.content;
    let metadata = response.data?.metadata;
    const token = response.data?.token;
    const hasContent = typeof content !== "undefined";
    const hasToken = typeof token !== "undefined";
    const hasMetadata = typeof metadata !== "undefined";

    if (
      messageHistory.length > 0 &&
      messageHistory[messageHistory.length - 1]["type"] !==
        ChatBotMessageType.Human
    ) {
      const lastMessage = messageHistory[messageHistory.length - 1];
      lastMessage.tokens = lastMessage.tokens || [];
      if (hasToken) {
        lastMessage.tokens.push(token);
      }

      lastMessage.tokens.sort((a, b) => a.sequenceNumber - b.sequenceNumber);
      if (lastMessage.tokens.length > 0) {
        const lastRunId =
          lastMessage.tokens[lastMessage.tokens.length - 1].runId;
        if (lastRunId) {
          lastMessage.tokens = lastMessage.tokens.filter(
            (c) => c.runId === lastRunId
          );
        }
      }

      if (!hasMetadata) {
        metadata = lastMessage.metadata;
      }

      if (hasContent) {
        chatSession.messageHistory = [...messageHistory.slice(0, messageHistory.length-1), 
          {
            ...lastMessage,
            type: ChatBotMessageType.AI,
            content,
            metadata,
            tokens: lastMessage.tokens,
          },
        ]
      } else {
        const contentFromTokens = lastMessage.tokens
          .map((c) => c.value)
          .join("");
          chatSession.messageHistory = [...messageHistory.slice(0, messageHistory.length-1), 
            {
              ...lastMessage,
              type: ChatBotMessageType.AI,
              content: contentFromTokens,
              metadata,
              tokens: lastMessage.tokens,
            },
          ]
      }
    } else {
      if (hasContent) {
        const tokens = hasToken ? [token] : [];
        chatSession.messageHistory = [...messageHistory, 
          {
            type: ChatBotMessageType.AI,
            content,
            metadata,
            tokens,
          },
        ]
      } else if (typeof token !== "undefined") {
        chatSession.messageHistory = [...messageHistory, 
          {
            type: ChatBotMessageType.AI,
            content: token.value,
            metadata,
            tokens: [token],
          },
        ]
      }
    }
  }
}

export async function getSignedUrl(key: string) {
  const signedUrl = await Storage.get(key as string);
  return signedUrl;
}

export function getSelectedModelMetadata(
  models: ModelItem[] | undefined,
  selectedModelOption: SelectProps.Option | null
): ModelItem | null {
  let selectedModelMetadata: ModelItem | null = null;

  if (selectedModelOption) {
    const { name, provider } = OptionsHelper.parseValue(
      selectedModelOption.value
    );
    const targetModel = models?.find(
      (m) => m.name === name && m.provider === provider
    );

    if (targetModel) {
      selectedModelMetadata = targetModel;
    }
  }

  return selectedModelMetadata;
}

