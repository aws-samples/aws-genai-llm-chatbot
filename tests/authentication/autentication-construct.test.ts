import { App, Stack } from "aws-cdk-lib";
import { Authentication } from "../../lib/authentication";
import { getTestConfig } from "../utils/config-util";
import { Template } from "aws-cdk-lib/assertions";

const app = new App();
const stack = new Stack(app);

new Authentication(stack, "AuthenticationConstruct", getTestConfig());

test("snapshot test", () => {
  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});
