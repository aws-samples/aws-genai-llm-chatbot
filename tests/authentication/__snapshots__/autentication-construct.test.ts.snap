// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
{
  "Outputs": {
    "AuthenticationConstructIdentityPoolIdE3D32668": {
      "Value": {
        "Ref": "AuthenticationConstructIdentityPool508C0E75",
      },
    },
    "AuthenticationConstructUserPoolId1F16F432": {
      "Value": {
        "Ref": "AuthenticationConstructUserPoolFE5ABE04",
      },
    },
    "AuthenticationConstructUserPoolLinkD2A7A02D": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "AWS::Region",
            },
            ".console.aws.amazon.com/cognito/v2/idp/user-pools/",
            {
              "Ref": "AuthenticationConstructUserPoolFE5ABE04",
            },
            "/users?region=",
            {
              "Ref": "AWS::Region",
            },
          ],
        ],
      },
    },
    "AuthenticationConstructUserPoolWebClientId09F05333": {
      "Value": {
        "Ref": "AuthenticationConstructUserPoolUserPoolClientDCD6FB5B",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AuthenticationConstructIdentityPool508C0E75": {
      "DependsOn": [
        "AuthenticationConstructUserPoolFE5ABE04",
        "AuthenticationConstructUserPoolsmsRole65D4F248",
        "AuthenticationConstructUserPoolUserPoolClientDCD6FB5B",
      ],
      "Properties": {
        "AllowUnauthenticatedIdentities": false,
        "CognitoIdentityProviders": [
          {
            "ClientId": {
              "Ref": "AuthenticationConstructUserPoolUserPoolClientDCD6FB5B",
            },
            "ProviderName": {
              "Fn::Join": [
                "",
                [
                  "cognito-idp.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".",
                  {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  {
                    "Ref": "AuthenticationConstructUserPoolFE5ABE04",
                  },
                ],
              ],
            },
            "ServerSideTokenCheck": true,
          },
        ],
      },
      "Type": "AWS::Cognito::IdentityPool",
    },
    "AuthenticationConstructIdentityPoolAuthenticatedRole8D31390C": {
      "DependsOn": [
        "AuthenticationConstructUserPoolFE5ABE04",
        "AuthenticationConstructUserPoolsmsRole65D4F248",
        "AuthenticationConstructUserPoolUserPoolClientDCD6FB5B",
      ],
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "ForAnyValue:StringLike": {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": {
                  "cognito-identity.amazonaws.com:aud": {
                    "Ref": "AuthenticationConstructIdentityPool508C0E75",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Default Authenticated Role for Identity Pool ",
              {
                "Fn::GetAtt": [
                  "AuthenticationConstructIdentityPool508C0E75",
                  "Name",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "AuthenticationConstructIdentityPoolDefaultRoleAttachmentDC1BB441": {
      "DependsOn": [
        "AuthenticationConstructUserPoolFE5ABE04",
        "AuthenticationConstructUserPoolsmsRole65D4F248",
        "AuthenticationConstructUserPoolUserPoolClientDCD6FB5B",
      ],
      "Properties": {
        "IdentityPoolId": {
          "Ref": "AuthenticationConstructIdentityPool508C0E75",
        },
        "Roles": {
          "authenticated": {
            "Fn::GetAtt": [
              "AuthenticationConstructIdentityPoolAuthenticatedRole8D31390C",
              "Arn",
            ],
          },
          "unauthenticated": {
            "Fn::GetAtt": [
              "AuthenticationConstructIdentityPoolUnauthenticatedRoleCB2B4EA2",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Cognito::IdentityPoolRoleAttachment",
    },
    "AuthenticationConstructIdentityPoolUnauthenticatedRoleCB2B4EA2": {
      "DependsOn": [
        "AuthenticationConstructUserPoolFE5ABE04",
        "AuthenticationConstructUserPoolsmsRole65D4F248",
        "AuthenticationConstructUserPoolUserPoolClientDCD6FB5B",
      ],
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "ForAnyValue:StringLike": {
                  "cognito-identity.amazonaws.com:amr": "unauthenticated",
                },
                "StringEquals": {
                  "cognito-identity.amazonaws.com:aud": {
                    "Ref": "AuthenticationConstructIdentityPool508C0E75",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Default Unauthenticated Role for Identity Pool ",
              {
                "Fn::GetAtt": [
                  "AuthenticationConstructIdentityPool508C0E75",
                  "Name",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "AuthenticationConstructUserPoolFE5ABE04": {
      "DeletionPolicy": "Delete",
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-COG1",
              "reason": "Default password policy requires min length of 8, digits, lowercase characters, symbols and uppercase characters: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.PasswordPolicy.html",
            },
            {
              "id": "AwsSolutions-COG2",
              "reason": "MFA not required for user usage.",
            },
          ],
        },
      },
      "Properties": {
        "AccountRecoverySetting": {
          "RecoveryMechanisms": [
            {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": {
          "AllowAdminCreateUserOnly": true,
        },
        "AutoVerifiedAttributes": [
          "email",
          "phone_number",
        ],
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "EnabledMfas": [
          "SMS_MFA",
        ],
        "MfaConfiguration": "OPTIONAL",
        "SmsConfiguration": {
          "ExternalId": "AuthenticationConstructUserPoolE66A108F",
          "SnsCallerArn": {
            "Fn::GetAtt": [
              "AuthenticationConstructUserPoolsmsRole65D4F248",
              "Arn",
            ],
          },
        },
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UserPoolAddOns": {
          "AdvancedSecurityMode": "ENFORCED",
        },
        "UsernameAttributes": [
          "email",
        ],
        "VerificationMessageTemplate": {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Delete",
    },
    "AuthenticationConstructUserPoolUserPoolClientDCD6FB5B": {
      "Properties": {
        "AllowedOAuthFlows": [
          "implicit",
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin",
        ],
        "CallbackURLs": [
          "https://example.com",
        ],
        "ExplicitAuthFlows": [
          "ALLOW_USER_PASSWORD_AUTH",
          "ALLOW_ADMIN_USER_PASSWORD_AUTH",
          "ALLOW_USER_SRP_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "GenerateSecret": false,
        "SupportedIdentityProviders": [
          "COGNITO",
        ],
        "UserPoolId": {
          "Ref": "AuthenticationConstructUserPoolFE5ABE04",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "AuthenticationConstructUserPoolsmsRole65D4F248": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Condition": {
                "StringEquals": {
                  "sts:ExternalId": "AuthenticationConstructUserPoolE66A108F",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cognito-idp.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "sns:Publish",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "sns-publish",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
